### PR TITLE
MNT: make extra_args in MovieWriter more forgiving

### PR DESCRIPTION
**Pull request by tacaswell**
As currently set up it assumes that there in as rcParams entry
to extract the default extra arguments, this makes that optional to
ease writing non-core MovieWriter sub-classes.

addresses part of #5403 

attn @WarrenWeckesser @dopplershift @WeatherGod

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
Could you also document `args_key` in the `MovieWriter` class docstring?

**Comment by mdboom** [comment]
`args_key` is something to be overwritten in a subclass?

**Comment by tacaswell** [comment]
Done and corrected docstrings a bit.

**Comment by WarrenWeckesser** [comment]
Thanks, that helps.

It seems that the existing use-cases of `MovieWriter` are as a wrapper of some command-line executable (hence `bin_path()` and `isAvailable()`).  The subclass I'm writing uses a Python library; there is no external command to be run.  That's easy enough to work around, but at the moment it is still not clear to me what *must* be implemented in a `MovieWriter` to ensure that an instance will work as the `writer` argument of the `save()` method of an `Animation`.  So, I'm off to read some more code... :smile:

**Comment by tacaswell** [comment]
Removed my attempt to update `exec_key`, there is multiple inheritance involved ....

**Comment by WarrenWeckesser** [comment]
>  it is still not clear to me what must be implemented in a MovieWriter to ensure that an instance will work as the writer argument of the save() method of an Animation.

I put a note about this at https://github.com/matplotlib/matplotlib/issues/5403#issuecomment-153950750

File: lib/matplotlib/animation.py [hunk 1]
```diff
@@ -123,7 +137,7 @@ def __init__(self, fps=5, codec=None, bitrate=None, extra_args=None,
         extra_args: list of strings or None
             A list of extra string arguments to be passed to the underlying
             movie utiltiy. The default is None, which passes the additional
-            argurments in the 'animation.extra_args' rcParam.
+            argurments in the `self.args_key` rcParam.
```
Comment (WeatherGod) [inline lib/matplotlib/animation.py]: `self.args_key` isn't an rcParam.

Comment (WeatherGod) [inline lib/matplotlib/animation.py]: also, could you fix the spelling mistake at the beginning of this line?

Comment (WeatherGod) [inline lib/matplotlib/animation.py]: btw, what I mean by my comment that it isn't an rcParam is that one could read it like that the rcParam is called "self.args_key", which doesn't make sense. This is going to need some rewording to be clearer.

**Comment by tacaswell** [comment]
@WeatherGod does this make it better or worse?

**Comment by WeatherGod** [comment]
+1 on the docstring improvement. Still a little bit unclear (might benefit from an example), but at least it isn't misleading.

**Comment by dopplershift** [comment]
Seems better to me as well.

**Comment by anntzer** [comment]
@dopplershift Could we just deprecate the various `animation.foo_args` rcparams?  Is there a specific use case for them?  (I can't think of one, but I have not thought *very* hard about this.)
If you want to maintain the ability to specify additional args, that functionality could be moved into the `animation.foo_path` rcparam, by allowing one to set it to `"path -arg -arg ..."` (parsed using shlex.split), which is not too bad to implement, but a wholesale removal of that functionality would be easier (note that I am not proposing to remove the `extra_args` kwarg to the writer classes themselves, just the rcparam).

**Comment by dopplershift** [comment]
I'm not passionate about keeping it, but I have no idea if anyone is using it. The idea is that you could put some special setting in matplotlibrc and not have to keep putting them in all your scripts.

**Comment by anntzer** [comment]
I'll look into merging it with `animation.foo_path` then (as the original issue here -- that third-party writers cannot have an associated rcparam and thus need to work around the entire machinery -- remains).

**Comment by anntzer** [comment]
Superseded by #16081 (and subsequent PRs).

**Pull request closed without merging**